Fix application:env and stoping/starting related issues

yfyf opened the pull request:
This seems to work nicely.

However, I am a bit unhappy about the fact that there's a cyclic dependency between `piqi_rpc` and `piqi_rpc_http` modules. The fact that piqi-rpc listeners live in a different supervision tree opens up some more possibilities for race conditions, when adding/removing services and/or crashing, so it would be good to rethink this from scratch, but I'm too braindead for the moment to do it, so just pushing this out with a hope that someone else will spot any obvious issues if they're out there.

yfyf committed:
Enlist all PHONY targets in Makefile

yfyf committed:
Don't start cowboy as part of piqi-rpc supervision tree

This fixes all kinds of nasty race conditions in stopping piqi-rpc.

Additionally, see [this](https://github.com/extend/cowboy/pull/265) issue on
why this is not supported and not the cowboy way of doing things.

Also added tests and refactored the `piqi_rpc_http:get_http_env` to not require
to always pass the default explicitly.

yfyf committed:
Move env maintenance using a gen_server / ETS

This a standard technique to maintain some stateful application configuration:
an ETS table owned by the supervisor + gen_server communicating with it.

This removes all the nasty hacks needed for cleaning the application env when
stoppin and guarantees that restarting the application will always give you a
fresh state.

Also, added some more tests.

stroop23 commented on src/piqi_rpc_http.erl:
```diff
@@ -15,25 +15,27 @@
-    cowboy:start_http(get_http_env(name, ?DEFAULT_NAME), get_http_env(nb_acceptors, 100),
-        [{port, get_http_env(port, ?DEFAULT_PORT)}],
-        [{env, [{dispatch, Routes}]}]
+start_listeners() ->
```
start_listener(), there's only one

stroop23 commented on src/piqi_rpc.erl:
```diff
@@ -88,18 +108,27 @@ remove_service(RpcServiceDef) ->
 
+-spec remove_all_services() -> ok.
+remove_all_services() ->
+    ok = piqi_rpc_http:cleanup(),
```
This will also stop the listener, i don't think we want to do that.

yfyf replied:
I changed `cleanup/0` to only unset the cowboy dispatching rules (and added a separate method for stopping the listener), so this is fixed now.

yfyf commented:
Ugh, guess what, the weird race conditions I was seeing turned out to be a bug in ~~cowboy~~ ranch:

```
23> [begin {ok, _} = cowboy:start_http(foo, 100, [{port, 8080}] , []), ok = cowboy:stop_listener(foo), io:format("~p~n", [N]) end || N <- lists:seq(1, 10) ].
1
2
3
4
5
6
** exception error: no match of right hand side value {error,{shutdown,{child,undefined,
                                                                              {ranch_listener_sup,foo},
                                                                              {ranch_listener_sup,start_link,
                                                                                                  [foo,100,ranch_tcp,[{port,8080}],cowboy_protocol,[]]},
                                                                              permanent,5000,supervisor,
                                                                              [ranch_listener_sup]}}}
```

yfyf committed:
Final touches on the start/stop problems

Figured out that the problem was a ranch bug.

Also, made sure piqi_rpc_http (i.e. cowboy) is always in sync with piqi_rpc
server even if it crashes and restarts.

yfyf commented:
ranch bug reported upstream: https://github.com/extend/ranch/issues/74
